src/System.Linq/src/System/Linq/Take.cs

```diff
@@ -96,5 +97,54 @@ private static IEnumerable<TSource> TakeWhileIterator<TSource>(IEnumerable<TSour
                 yield return element;
             }
         }
+
+        public static IEnumerable<TSource> TakeLast<TSource>(this IEnumerable<TSource> source, int count)
+        {
+            if (source == null)
+            {
+                throw Error.ArgumentNull(nameof(source));
+            }
+
+            if (count <= 0)
+            {
+                return EmptyPartition<TSource>.Instance;
+            }
+
+            return TakeLastIterator(source, count);
+        }
+
+        private static IEnumerable<TSource> TakeLastIterator<TSource>(IEnumerable<TSource> source, int count)
+        {
+            Debug.Assert(source != null);
+            Debug.Assert(count > 0);
+
+            var queue = new Queue<TSource>();
+
+            using (IEnumerator<TSource> e = source.GetEnumerator())
+            {
+                while (e.MoveNext())
+                {
+                    if (queue.Count < count)
+                    {
+                        queue.Enqueue(e.Current);
+                    }
+                    else
+                    {
+                        do
+                        {
+                            queue.Dequeue();
+                            queue.Enqueue(e.Current);
+                        }
+                        while (e.MoveNext());
+                        break;
+                    }
+                }
+            }
+
+            while (queue.Count > 0)
+            {
+                yield return queue.Dequeue();
+            }
```

Comment:
(The class idea shifts the point of obtaining results from the first `MoveNext()` to the `GetEnumerator()` which differs from other methods, though perhaps acceptable since it doesn't break anything being new. Otherwise I'd still wonder about `foreach`ing the queue rather than removing from it).